core/src/main/java/com/graphhopper/reader/osm/OSMReader.java

```diff
@@ -596,21 +599,45 @@ static List<OSMTurnRestriction> createTurnRelations(ReaderRelation relation, Str
+        	OSMTurnRestriction.ViaType viaType = OSMTurnRestriction.ViaType.UNSUPPORTED;
             // we use -1 to indicate 'missing', which is fine because we exclude negative OSM IDs (see #2652)
-            long viaNodeID = -1;
             long toWayID = -1;
-
+            
+            // A Turn Restriction (https://wiki.openstreetmap.org/wiki/Relation:restriction) can consist of...
+            
             for (ReaderRelation.Member member : relation.getMembers()) {
-                if (ReaderElement.Type.WAY == member.getType() && "to".equals(member.getRole()))
+                // ONE TO Way (a "no_exit" restriction can have ONE OR MORE TO ways)
+                if (ReaderElement.Type.WAY == member.getType() && "to".equals(member.getRole())) {
                     toWayID = member.getRef();
-                else if (ReaderElement.Type.NODE == member.getType() && "via".equals(member.getRole()))
-                    viaNodeID = member.getRef();
+                }
+                
+               // ONE VIA NODE or
+                else if (ReaderElement.Type.NODE == member.getType() && "via".equals(member.getRole())) {
+                	if (viaType == OSMTurnRestriction.ViaType.UNSUPPORTED) {
+                		viaIDs.add(member.getRef());
+                		viaType = OSMTurnRestriction.ViaType.VIA_NODE;
+                	}
+                }
+                // ONE OR MORE VIA WAYS
+                else if (ReaderElement.Type.WAY == member.getType() && "via".equals(member.getRole())) {
+                	if (viaType == OSMTurnRestriction.ViaType.UNSUPPORTED) {
+                		viaIDs.add(member.getRef());
+                		viaType = OSMTurnRestriction.ViaType.VIA_WAY;
+                	} 
+                	else if (viaType == OSMTurnRestriction.ViaType.VIA_WAY || viaType == OSMTurnRestriction.ViaType.MULTI_VIA_WAY) {
+                		viaIDs.add(member.getRef());
+                		viaType = OSMTurnRestriction.ViaType.MULTI_VIA_WAY;
+                	}
+                }
             }
-            if (toWayID >= 0 && viaNodeID >= 0) {
+            
+            // ONE FROM Way (a "no_entry" restriction can have ONE OR MORE FROM ways)
+            if (toWayID >= 0 && viaIDs.size() > 0) {
                 List<OSMTurnRestriction> res = new ArrayList<>(2);
                 for (ReaderRelation.Member member : relation.getMembers()) {
                     if (ReaderElement.Type.WAY == member.getType() && "from".equals(member.getRole())) {
-                        OSMTurnRestriction osmTurnRelation = new OSMTurnRestriction(member.getRef(), viaNodeID, toWayID, type);
+                        OSMTurnRestriction osmTurnRelation = new OSMTurnRestriction(member.getRef(), viaIDs, toWayID, type, viaType);
                         osmTurnRelation.setVehicleTypeRestricted(vehicleTypeRestricted);
                         osmTurnRelation.setVehicleTypesExcept(vehicleTypesExcept);
                         res.add(osmTurnRelation);
```

Comment:
@karussell in https://github.com/graphhopper/graphhopper/pull/2648 you implemented this functionality. 
1. Is it intentional to restrict the possible OSMTurnRestriction List to 2 elements?
2. Should we add a check that the restriction is a "no_entry"?
3. Should we extend this approach to TO ways since there multiple ways for "no_exit" are possible as well?

Reply:
> Do you mean the constructor argument of ArrayList? It is not restricted to 2 elements but just reduces the default from 10 but probably not necessary.

A I didn't know that its possible to specify the initialCapacity. I thought it limits the array 👍